### PR TITLE
Add applications

### LINKED ISSUE #59: Add Applications

### PR DESCRIPTION
**Pull request by jimouris**
Closes #59

### CODE REVIEW

File: draft-mouris-cfrg-mastic.md [hunk 1]
```diff
@@ -661,15 +680,154 @@ malicious Aggregator.
 
 > NOTE to be specified in full detail.
 
+# Applications of `Mastic`
```
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:683]:
```suggestion
# Motivating Applications
```

File: draft-mouris-cfrg-mastic.md [hunk 2]
```diff
@@ -661,15 +680,154 @@ malicious Aggregator.
...
+
+Based on the two modes of operation of Mastic (i.e., Weighted Heavy-Hitters
+{{weighted-heavy-hitters}} and Attribute-Based Metrics
+{{aggregation-by-labels}}), we describe two applications of interest.
+
+## Network Error Logging {#NEL}
```
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:689]:
```suggestion
### Network Error Logging {#NEL}

```

File: draft-mouris-cfrg-mastic.md (hunk 1 above)
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:683]: Let's make this a subsection of the intro.

File: draft-mouris-cfrg-mastic.md [hunk 3]
```diff
@@ -661,15 +680,154 @@ malicious Aggregator.
...
+
+NEL data is privacy-sensitive for two reasons. First, it exposes information
+that the server would not otherwise have access to, which can be abused to probe
+the client's network configuration as described in [{{W3C23}}, Section 9].
+Second, for operational reasons, the reporting endpoint may be organizationally
+separated from the server (i.e., run on different cloud infrastructures),
+leading to an increased risk of the client's browsing history being exposed
+(e.g., in a data breach).
+
+**Private NEL with Mastic.** MPC helps mitigate these risks by revealing to the
```
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:722]: **This doesn't render the way you want it to in all version of internet drafts**.

```suggestion
MPC helps mitigate these risks by revealing to the
```

File: draft-mouris-cfrg-mastic.md [hunk 4]
```diff
@@ -661,15 +680,154 @@ malicious Aggregator.
...
+of these errors are visible to the server, but not all: failures in DNS, TCP,
+TLS, and HTTP can occur without the server having any visibility into the issue.
+A small amount of connection errors is expected, even under normal operating
+conditions; but a sudden, substantial increase in errors may be an indication of
+an outage, or a configuration issue impacting millions of users. Without a
+reporting mechanism like NEL, these events would only manifest in the server's
+telemetry as a drop in overall traffic.
+
+NEL is particularly important for content delivery networks, such as Akamai,
+AWS, or Cloudflare, that handle HTTP traffic for a large number of websites
```
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:701]: Internet-Drafts shouldn't reference specific companies or institutions. "Content delivery network" is clear enough to the intended audience.

```suggestion
NEL is particularly important for content delivery networks that handle HTTP traffic for a large number of websites
```

File: draft-mouris-cfrg-mastic.md [hunk 5]
```diff
@@ -661,15 +680,154 @@ malicious Aggregator.
...
+reporting time), user agent (browser type and version), etc. However, our main
+goal is to understand 1) the distribution of errors and 2) which domains are
+impacted.
+
+We expect there to be a large number of distinct domain names (millions in the
+case of content delivery networks) and only a small number of error variants
+(the NEL spec {{W3C23}} defines 30 variants). The following Mastic parameters
+are suitable for this application.
+
+* **Inputs:** Each input `alpha` encodes the domain `dom` truncated to `n = 256`
```
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:743]:
```suggestion
* Each input `alpha` encodes the domain `dom` truncated to `n = 256`
```

File: draft-mouris-cfrg-mastic.md [hunk 6]
```diff
@@ -661,15 +680,154 @@ malicious Aggregator.
...
+bits, which is sufficient to represent most of the domains on the internet
+{{MST24}}, {{BBCGGI21}}. Domains that are shorter than `n` bits are padded with
+0s.
+* **Weights:** Each weight `beta` represents the error variant `dom`. To compute
```
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:747]:
```suggestion
0s.

* Each weight `beta` represents the error variant `dom`. To compute
```

File: draft-mouris-cfrg-mastic.md [hunk 7]
```diff
@@ -661,15 +680,154 @@ malicious Aggregator.
...
+the distribution of errors, we encode each error variant as a distinct bucket of
+a histogram so that `[1, 0, 0, ...]` represents `ok`, `[0, 1, 0, ...]`
+represents `dns.unreachable`, `[0, 0, 1, ...]` represents
+`dns.name_not_resolved`, and so on. There are 30 such variants (see {{W3C23}},
+Section 6), so the language `L` of valid weights is exactly the set of length-30
+vectors over `F` containing all 0s except for a single 1.
+* **Ordering:** Our `order(·)` function computes the ratio of reports with `err
```
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:754]: Note that unicode charcters, like "·", should not appear in internet drafts. With the exception of authors, it's a good idea to make sure every character is ASCII.
```suggestion
vectors over `F` containing all 0s except for a single 1.

* Our `order` function computes the ratio of reports with `err
```

Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:754]: Make sure we have defined `order` for the reader, probably in the conventions section.

File: draft-mouris-cfrg-mastic.md [hunk 8]
```diff
@@ -661,15 +680,154 @@ malicious Aggregator.
...
+domains but not others). Another benefit is that, under normal operating
+conditions, there will be a small number of heavy-hitting domains, which means
+Mastic will run very efficiently. During an incident, there will be more heavy
+hitters, which means it will take longer to compute the set of impacted domains.
+However, we get the errors immediately at the root of the prefix tree, which is
+the most important information needed to begin remediation. As more levels are
+evaluated we get more detailed errors.
+
+
+## Attribute-Based Browser Telemetry {#attribute-based-telemetry}
```
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:769]:
```suggestion
## Attribute-Based Browser Telemetry {#attribute-based-telemetry}

```

File: draft-mouris-cfrg-mastic.md [hunk 9]
```diff
@@ -661,15 +680,154 @@ malicious Aggregator.
...
+Web browsers, like Chrome, Firefox, or Safari, collect telemetry generated by
```
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:770]:
```suggestion
Web browsers collect telemetry generated by
```

File: draft-mouris-cfrg-mastic.md [hunk 10]
```diff
@@ -661,15 +680,154 @@ malicious Aggregator.
...
+users as they surf the web to gain insights into trends that guide product
+decisions. In many cases, Prio can be used to privately aggregate this
+telemetry. However, this comes at the cost of flexibility.
+
+For example, Mozilla is using Prio to collect page load metrics from Firefox for
+a list of known popular sites (e.g., `google.com`). The purpose of these metrics
```
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:776]:
```suggestion
a list of known popular sites (e.g., `example.com`). The purpose of these metrics
```

File: draft-mouris-cfrg-mastic.md [hunk 11]
```diff
@@ -661,15 +680,154 @@ malicious Aggregator.
...
+evaluated we get more detailed errors.
+
+
+## Attribute-Based Browser Telemetry {#attribute-based-telemetry}
+Web browsers, like Chrome, Firefox, or Safari, collect telemetry generated by
+users as they surf the web to gain insights into trends that guide product
+decisions. In many cases, Prio can be used to privately aggregate this
+telemetry. However, this comes at the cost of flexibility.
+
+For example, Mozilla is using Prio to collect page load metrics from Firefox for
```
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:775]: We'll need to remove mention Mozilla by name and speak about a hypothetical browser.

File: draft-mouris-cfrg-mastic.md [hunk 12]
```diff
@@ -661,15 +680,154 @@ malicious Aggregator.
...
+a list of known popular sites (e.g., `google.com`). The purpose of these metrics
+is to detect if changes to these sites cause regressions that might be
+correlated with an increased average load time or error rate. A subtle, but
+important requirement for this system is the ability to break down the metrics
+by client attributes. The most crucial attributes are 1) the software version,
+and 2) the information about the client's location.
```
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:781]: This is Mozilla's preference, but others might do something else.
```suggestion
by client attributes. Suppose for example that we want to aggregate by 1) the software version,
and 2) information about the client's location.
```

File: draft-mouris-cfrg-mastic.md [hunk 13]
```diff
@@ -661,15 +680,154 @@ malicious Aggregator.
...
+
+As we mentioned above, meeting this requirement by increasing the size of the
+histogram leads to intolerable communication overhead. An alternative is to have
```
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:784]: Have we actually mentioned this? We did in the paper, but I'm not sure we made the Prio3 comparison here.

FWIW, I don't think we necessarily need to. could we just remove this paragraph without interrupting the flow?

File: draft-mouris-cfrg-mastic.md [hunk 14]
```diff
@@ -661,15 +680,154 @@ malicious Aggregator.
...
+each client upload this information in the clear alongside its Prio report so
+that the reports can be grouped by version and location. The downside of this
+approach is that it significantly reduces the anonymity set of each user since
+they are only mixed with their attribute group rather than the entire
+population.
+
+**Private Browser Telemetry with Mastic.** Mastic provides a simple solution to
```
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:791]:
```suggestion
Mastic provides a simple solution to
```

File: draft-mouris-cfrg-mastic.md [hunk 15]
```diff
@@ -661,15 +680,154 @@ malicious Aggregator.
...
+this problem. For the sake of presentation, we consider a simplified version of
+Mozilla's use case (the same approach can be applied to any aggregation task for
+which Prio is suitable). Each client reports a tuple `(ver, loc, site, time)`
```
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:794]:
```suggestion
which Prio3 ({{Section 7 of !VDAF}}) is suitable). Each client reports a tuple `(ver, loc, site, time)`
```

File: draft-mouris-cfrg-mastic.md [hunk 16]
```diff
@@ -661,15 +680,154 @@ malicious Aggregator.
...
+where: `ver` is a string representing the client's software version (e.g.,
+`Firefox/122.0`); `loc` is a string encoding its country code (e.g., `GR`, `US`,
```
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:796]:
```suggestion
where: `ver` is a string representing the client's software version; `loc` is a string encoding its country code (e.g., `GR`, `US`,
```

File: draft-mouris-cfrg-mastic.md [hunk 17]
```diff
@@ -661,15 +680,154 @@ malicious Aggregator.
...
+`IN`, etc.); `site` is one of a fixed set of sites (e.g., `google.com`,
```
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:797]:
```suggestion
`IN`, etc.); `site` is one of a fixed set of sites (e.g., `example.com`,
```

File: draft-mouris-cfrg-mastic.md [hunk 18]
```diff
@@ -661,15 +680,154 @@ malicious Aggregator.
...
+inputs here is relatively small: there are less than 200 country codes and a
+handful of browser versions in wide use at any given time. This means the
+aggregators can enumerate a set of inputs of interest and evaluate them
+immediately. Consider the following parameters for Mastic, in its
+attribute-based metrics mode of operation {{aggregation-by-labels}}:
+* **Attributes:** Two-letter country codes can easily be encoded in 2 bytes.
+Likewise, the number of distinct browser versions is easily less than 216, so 2
+bytes are sufficient. Therefore, each `alpha` can be encoded with just `n =
+32`bits.
+* **Values:** Similar to private NEL, each weight `beta` is a 0-vector except
```
Comment (cjpatton) [inline draft-mouris-cfrg-mastic.md:813]:
```suggestion

* Similar to private NEL, each weight `beta` is a 0-vector except
```

File: draft-mouris-cfrg-mastic.md (hunk 13 above)
Comment (jimouris) [reply draft-mouris-cfrg-mastic.md:784]: good catch.

**Review by cjpatton — APPROVED** [review]

**Pull request merged**